src/ui/articlesaver.cc

```diff
@@ -0,0 +1,261 @@
+      QRegularExpression anchorRx( "(g[0-9a-f]{32}_[0-9a-f]+_)" );
+      auto match = rx3.match( html, pos );
+      while ( match.hasMatch() ) {
+        pos          = match.capturedStart();
+        QString name = QUrl::fromPercentEncoding( match.captured( 2 ).simplified().toLatin1() );
+        QString anchor;
+        name.replace( "?gdanchor=", "#" );
+        int n = name.indexOf( '#' );
+        if ( n > 0 ) {
+          anchor = name.mid( n );
+          name.truncate( n );
+          anchor.replace( anchorRx, "\\1" );
+        }
+        name.replace( rxName, "_" );
+        name = QString( R"(href=")" ) + QUrl::toPercentEncoding( name ) + ".html" + anchor + "\"";
+        html.replace( pos, match.captured().length(), name );
+        pos += name.length();
+        match = rx3.match( html, pos );
+      }
+
+      // MDict anchors
+      QRegularExpression anchorLinkRe(
+        R"((<\s*a\s+[^>]*\b(?:name|id)\b\s*=\s*["']*g[0-9a-f]{32}_)([0-9a-f]+_)(?=[^"']))",
+        QRegularExpression::PatternOption::CaseInsensitiveOption | QRegularExpression::UseUnicodePropertiesOption );
+      html.replace( anchorLinkRe, "\\1" );
+
+      if ( complete ) {
+        QString folder = fi.absoluteDir().absolutePath() + "/" + fi.baseName() + "_files";
+        static QRegularExpression rx1( R"lit("((?:bres|gico|gdau|qrcx|qrc|gdvideo)://[^"]+)")lit" );
+        static QRegularExpression rx2( "'((?:bres|gico|gdau|qrcx|qrc|gdvideo)://[^']+)'" );
+        set< QByteArray > resourceIncluded;
+        vector< pair< QUrl, QString > > downloadResources;
+
+        filterAndCollectResources( html, rx1, "\"", folder, resourceIncluded, downloadResources );
+        filterAndCollectResources( html, rx2, "'", folder, resourceIncluded, downloadResources );
+
+        auto * progressDialog = new ArticleSaveProgressDialog( uiParent_ );
+        int maxVal = 1; // main html
+        bool anyHandlerConnected = false;
+
+        for ( const auto & p : downloadResources ) {
+          ResourceToSaveHandler * handler = view_->saveResource( p.first, p.second );
+          if ( handler && !handler->isEmpty() ) {
+            anyHandlerConnected = true;
+            maxVal += 1;
+            QObject::connect( handler,
+                  &ResourceToSaveHandler::done,
+                  progressDialog,
+                  &ArticleSaveProgressDialog::perform );
+          }
+        }
+
+        progressDialog->setLabelText( QObject::tr( "Saving article..." ) );
+        progressDialog->setRange( 0, maxVal );
+
+        if ( !anyHandlerConnected ) {
+          // No resource handlers attached — no background work expected.
+          // Close and schedule deletion of the dialog immediately to avoid
+          // leaving it open in rare edge-cases.
+          progressDialog->setValue( maxVal );
+          progressDialog->close();
+          progressDialog->deleteLater();
+        }
+        else {
+          progressDialog->setValue( 0 );
+          progressDialog->show();
```

Comment:
if there is no resources needed to be download ,the dialog has no need to show .

the `else` branch can be removed.